File: 2020/24/Rust/src/main.rs
use std::collections::HashMap;
use std::env;
use std::fs;

const ITERATIONS: usize = 100;

type Point = (i32, i32);
type Grid = HashMap<Point, Tile>;

#[derive(Debug)]
enum Dir {
    East,
    SouthEast,
    SouthWest,
    West,
    NorthWest,
    NorthEast,
}

impl Dir {
    fn gen_dir(first: &char, second: &char) -> Dir {
        let direction = vec![first, second];
        if direction == vec![&'s', &'e'] {
            Dir::SouthEast
        } else if direction == vec![&'s', &'w'] {
            Dir::SouthWest
        } else if direction == vec![&'n', &'w'] {
            Dir::NorthWest
        } else if direction == vec![&'n', &'e'] {
            Dir::NorthEast
        } else {
            panic!("Invalid direction");
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq)]
enum Tile {
    Black,
    White,
}

impl Tile {
    fn flip(&self) -> Tile {
        match self {
            Tile::Black => Tile::White,
            Tile::White => Tile::Black,
        }
    }
}

fn parse_line(line: &str) -> Vec<Dir> {
    let line = line.chars().collect::<Vec<_>>();
    let mut result = Vec::new();
    let mut i = 0;
    while i < line.len() {
        let first = line[i];
        if first == 'e' {
            result.push(Dir::East);
        } else if first == 'w' {
            result.push(Dir::West);
        } else {
            i += 1;
            let second = line[i];
            let direction = Dir::gen_dir(&first, &second);
            result.push(direction);
        }
        i += 1;
    }
    result
}

fn parse_input(input: &str) -> Vec<Vec<Dir>> {
    input
        .lines()
        .filter(|x| !x.is_empty())
        .map(|x| parse_line(x))
        .collect()
}

fn grid_lookup(point: &Point, grid: &Grid) -> Tile {
    if let Some(tile) = grid.get(point) {
        *tile
    } else {
        Tile::White
    }
}

fn step_grid(point: &Point, direction: &Dir) -> Point {
    let (x, y) = point;
    match direction {
        Dir::East => (x + 2, *y),
        Dir::NorthEast => (x + 1, y + 1),
        Dir::SouthEast => (x + 1, y - 1),
        Dir::West => (x - 2, *y),
        Dir::NorthWest => (x - 1, y + 1),
        Dir::SouthWest => (x - 1, y - 1),
    }
}

fn find_tile(tile: &[Dir]) -> Point {
    let mut point = (0, 0);
    for entry in tile {
        point = step_grid(&point, entry);
    }
    point
}

fn step_input(tiles: &[Vec<Dir>]) -> Grid {
    let mut grid: Grid = HashMap::new();
    for tile in tiles {
        let point = find_tile(tile);
        let next = grid_lookup(&point, &grid).flip();
        grid.insert(point, next);
    }
    grid
}

fn get_adjacent(point: &Point) -> Vec<Point> {
    let x = point.0;
    let y = point.1;
    let adjacent = |t: i32| (t - 1)..=(t + 1);
    let mut result = Vec::new();
    for a in adjacent(x) {
        for b in adjacent(y) {
            if a != x && b != y {
                result.push((a, b))
            }
        }
    }
    result.push((x - 2, y));
    result.push((x + 2, y));
    result
}

fn next_tile(point: &Point, grid: &Grid) -> Tile {
    let tile = grid_lookup(point, grid);
    let adjacent = get_adjacent(point);
    let count = adjacent
        .iter()
        .filter(|x| grid_lookup(x, grid) == Tile::Black)
        .count();
    if tile == Tile::White && count == 2 {
        Tile::Black
    } else if tile == Tile::Black && (count == 0 || count > 2) {
        Tile::White
    } else {
        tile
    }
}

fn update_tiles(grid: &Grid) -> Grid {
    let mut next_grid = HashMap::new();
    let points = grid.keys();
    for point in points {
        let next = next_tile(point, grid);
        next_grid.insert(*point, next);
    }
    next_grid
}

fn cycle_tiles(grid: &Grid) -> Grid {
    let mut grid = grid.to_owned();
    for _ in 0..ITERATIONS {
        let keys = grid
            .iter()
            .filter(|(_, x)| x == &&Tile::Black)
            .map(|(k, _)| k.to_owned());
        let not_included = keys
            .flat_map(|x| get_adjacent(&x))
            .filter(|x| !grid.contains_key(x))
            .map(|x| (x, Tile::White))
            .collect::<HashMap<_, _>>();
        grid.extend(not_included);
        grid = update_tiles(&grid);
    }
    grid
}

fn solve(input: &[Vec<Dir>]) -> (usize, usize) {
    let count_black_tiles =
        |grid: &Grid| grid.values().filter(|x| x == &&Tile::Black).count();

    let stepped = step_input(&input);
    let part_1 = count_black_tiles(&stepped);

    let cycled = cycle_tiles(&stepped);
    let part_2 = count_black_tiles(&cycled);

    (part_1, part_2)
}

fn output(path: &str) {
    let input = fs::read_to_string(&path).unwrap();
    let parsed = parse_input(&input);

    let (part_1, part_2) = solve(&parsed);

    println!("File: {}", path);
    println!("  Part 1: {}", part_1);
    println!("  Part 2: {}", part_2);
}

fn main() {
    env::args().skip(1).for_each(|x| output(&x));
}

File: 2020/24/Haskell/src/Main.hs
import           Data.HashMap.Strict            ( HashMap )
import qualified Data.HashMap.Strict           as HM
import           Data.List
import           System.Environment
import           Text.Printf

type Point = (Int, Int)
type TileGrid = HashMap Point Tile

data Tile = Black | White deriving Eq

flipTile :: Tile -> Tile
flipTile Black = White
flipTile White = Black

parseLine :: String -> [String]
parseLine []          = []
parseLine ('e'   : x) = "e" : parseLine x
parseLine ('w'   : x) = "w" : parseLine x
parseLine (a : b : x) = [a, b] : parseLine x
parseLine _           = []

stepGrid :: Point -> String -> Point
stepGrid (x, y) "e"  = (x + 2, y)
stepGrid (x, y) "ne" = (x + 1, y + 1)
stepGrid (x, y) "se" = (x + 1, y - 1)
stepGrid (x, y) "w"  = (x - 2, y)
stepGrid (x, y) "nw" = (x - 1, y + 1)
stepGrid (x, y) "sw" = (x - 1, y - 1)
stepGrid _      _    = (0, 0)

findTile :: Point -> [String] -> Point
findTile p []       = p
findTile p (x : xs) = findTile p' xs where p' = stepGrid p x

stepInput :: TileGrid -> [String] -> TileGrid
stepInput t []        = t
stepInput t ([] : xs) = stepInput t xs
stepInput t (x  : xs) = stepInput t' xs
 where
  parsed = parseLine x
  tile   = findTile (0, 0) parsed
  value  = HM.findWithDefault White tile t
  t'     = HM.insert tile (flipTile value) t

getAdjacent :: Point -> [Point]
getAdjacent (x, y) = diagonals ++ line
 where
  diagonals =
    [ (a, b) | a <- [x - 1 .. x + 1], a /= x, b <- [y - 1 .. y + 1], b /= y ]
  line = [(x - 2, y), (x + 2, y)]

nextTile :: Point -> TileGrid -> Tile
nextTile p t = tile
 where
  value    = HM.findWithDefault White p t
  adjacent = getAdjacent p
  count    = length . filter (\x -> HM.lookup x t == Just Black) $ adjacent
  tile | value == White && count == 2 = Black
       | value == Black && (count == 0 || count > 2) = White
       | otherwise                    = value

updateTiles :: TileGrid -> [Point] -> TileGrid -> TileGrid
updateTiles t [] t0 | HM.null t = updateTiles t (HM.keys t0) t0
                    | otherwise = t
updateTiles t (x : xs) t0 = updateTiles t' xs t0
 where
  tile = nextTile x t0
  t'   = HM.insert x tile t

cycleTiles :: Int -> Int -> TileGrid -> TileGrid
cycleTiles i0 i t | i == i0   = t
                  | otherwise = cycleTiles i0 (i + 1) next
 where
  keys = HM.keys . HM.filter (== Black) $ t
  notIncluded =
    map head
      . group
      . sort
      . filter (not . (`HM.member` t))
      . concatMap getAdjacent
      $ keys
  n    = HM.fromList . zip notIncluded $ repeat White
  t'   = HM.union t n
  next = updateTiles HM.empty [] t'

part1 :: [String] -> Int
part1 = length . filter (== Black) . HM.elems . stepInput HM.empty

part2 :: [String] -> Int
part2 =
  length . filter (== Black) . HM.elems . cycleTiles 100 0 . stepInput HM.empty

output :: String -> IO ()
output path = do
  input <- lines <$> readFile path
  printf "File: %s\n" path
  printf "  Part 1: %d\n" . part1 $ input
  printf "  Part 2: %d\n" . part2 $ input

main :: IO ()
main = getArgs >>= mapM_ output
